video/src/ai/eto/rikai/sql/spark/datasources/VideoPartitionReaderFactory.scala

```diff
@@ -98,40 +63,64 @@ case class VideoPartitionReaderFactory(
       case _ =>
         file.filePath
     }
+  }
 
-    val grabber = new FFmpegFrameGrabber(tmpPath)
+  private def buildGrabber(
+      options: VideoOptions,
+      file: PartitionedFile
+  ): FFmpegFrameGrabber = {
+    val path = resolveFilePath(file)
+    val grabber = new FFmpegFrameGrabber(path)
     grabber.setImageWidth(options.imageWidth)
     grabber.setImageHeight(options.imageHeight)
-    grabber.setImageScalingFlags(getImageScalingFlags(options.scalerFlag))
-
-    val converter = new Java2DFrameConverter
+    grabber.setImageScalingFlags(options.getImageScalingFlags())
     grabber.start()
-    val numOfFps = getNumberOfFps(grabber)
     grabber.setFrameNumber(file.start.toInt)
+    grabber
+  }
+
+  override def buildReader(
+      file: PartitionedFile
+  ): PartitionReader[InternalRow] = {
+    val uri = new URI(file.filePath)
+    val grabber = buildGrabber(options, file)
+    val (frameStep, offset) = options.getFrameStep(grabber)
+
     var frame: Frame = null
-    var frame_id = file.start
+    var frameId = file.start - frameStep + offset
 
-    new PartitionReader[InternalRow] {
+    val converter = new Java2DFrameConverter
+
+    val videoReader = new PartitionReader[InternalRow] {
       override def next(): Boolean = {
-        val targetFrameId = frame_id + numOfFps
-        while (frame_id < targetFrameId - 1) {
-          frame_id = frame_id + 1
+        val targetFrameId = frameId + frameStep
+        while (frameId < targetFrameId - 1) {
+          frameId = frameId + 1
           grabber.grabImage()
```

Comment:
I saw `grabImage()` here equals to `grabFrame(false, true, true, false, false)`, will change some of this parameter make it more efficient?

Reply:
I don't know yet. Let us try in another performance tuning pull request.